### ISSUE TITLE
Connect MapNode with MapNode bug

### ISSUE THREAD

**Issue by anbai106** [nipy/nipype#2052]
### Summary
I tried to connect some MapNode to parallelize my pipeline, but I found maybe a bug, I know that the input for a MapNode is a list containing some string, so the next MapNode will process this string in parallel, to be clear, I will give you a specific example

This is a part of dwi preprocessing pipeline, I try to parallelize the pipeline.
I have two subjects: subject_01, subject_02,

at some point, I have two MapNodes:
`split_moving = pe.MapNode(fsl.Split(dimension='t'), name='split_b0_moving', iterfield=['in_file'])
    flirt = pe.MapNode(fsl.FLIRT(interp='spline', dof=6, bins=50, save_log=True, cost='corratio', cost_func='corratio', padding_size=10,
                                 searchr_x=[-4, 4], searchr_y=[-4, 4], searchr_z=[-4, 4], fine_search=1, coarse_search=10),
                       name='b0_co_registration', iterfield=['reference', 'in_weight', 'ref_weight'])`
Firstly, I split the dwi images, so the output of this MapNode will be **a list containing two lists**:
`[['/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0000.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0001.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0002.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0003.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0004.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0005.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0006.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0007.nii.gz'], ['/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0000.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0001.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0002.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0003.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0004.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0005.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0006.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0007.nii.gz']]`

And I pass this to the **in_file** of flirt MapNode, but it seems that this does not work, errors came back:
`TraitError: The 'in_file' trait of a FLIRTInputSpec instance must be an existing file name, but a value of [['/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0000.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0001.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0002.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0003.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0004.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0005.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0006.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0007.nii.gz'], ['/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0000.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0001.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0002.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0003.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0004.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0005.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0006.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0007.nii.gz']] <type 'list'> was specified.
`

I dont know how to deal with this situation, because the output of fsl.Split will be a list of split images, and MapNode will create a new list containing the resulting lists of the two subjects.

Do you have any idea what I should do with this?

Is this a bug???

Thanks in advance
 

### Platform details:

Please paste the output of: `python -c "import nipype; print(nipype.get_info()); print(nipype.__version__)"`

### Execution environment

Choose one
- Container [Tag: ???]
- My python environment inside container [Base Tag: ???]
- My python environment outside container

**Comment by effigies** [comment]
You have two options.

1) If you want to preserve the shape of these lists, you'll need to add `nested=True` to the second `MapNode`'s arguments.

2) If you're okay with your lists being merged into a single list, you can do:

```Python
merge = pe.Node(niu.Merge(), name='merge')
wf.connect([
    (split_moving, merge, [('out_file', 'in1')]),
    (merge, flirt, [('out', 'in_file')])])
```

(I haven't looked up the actual names, so you may need to adjust `out_file` and `out` in those connections.)

Merge will give you a flattened list, and everything going forward will be based on that. The nested approach might be better if you need to keep the lists separate for the future, which it sounds like you might, if you're iterating over subjects with `MapNode`s.

**Comment by anbai106** [comment]
@effigies  I think I will try to use the option one, because I dont wanna a flattened list which will be regarded as many subjects as the length of the list by MapNode, actually, for flirt, there are other MapNode to connect as inputs, e.g the mask from Bet, which are always a list of two files.

I will give you the feedback

Thanks anyway

**Comment by anbai106** [comment]
@effigies Hi, I tired the first solution, but it seems that this flag does not really work, I still get the same errors:
`    fslroi_ref = pe.MapNode(fsl.ExtractROI(args='0 1'), name='b0_reference', iterfield=['in_file'], nested=True)
    tsize = nb_b0 - 1
    fslroi_moving = pe.MapNode(fsl.ExtractROI(args='1 '+str(tsize)), name='b0_moving', iterfield=['in_file'])
    split_moving = pe.MapNode(fsl.Split(dimension='t'), name='split_b0_moving', iterfield=['in_file'], nested=True)

    bet_ref = pe.MapNode(fsl.BET(frac=0.3, mask=True, robust=True),
                       name='bet_ref', iterfield=['in_file'])

    dilate = pe.MapNode(fsl.maths.MathsCommand(nan2zeros=True,
                     args='-kernel sphere 5 -dilM'), name='mask_dilate', iterfield=['in_file'], nested=True)

    # list_4_flirt = pe.MapNode(niu.Function(input_names=['in_files'],
    #                                        output_names=['out_file'],
    #                                        function=convert_listoflist_to_list),
    #                           name='listoflisttolist', iterfield=['in_files'])

    flirt = pe.MapNode(interface=fsl.FLIRT(interp='spline', dof=6, bins=50, save_log=True, cost='corratio', cost_func='corratio', padding_size=10,
                                 searchr_x=[-4, 4], searchr_y=[-4, 4], searchr_z=[-4, 4], fine_search=1, coarse_search=10),
                       name='b0_co_registration',
                       iterfield=['reference', 'in_weight', 'ref_weight'],

### Connect MapNodes
 wf.connect([
        (inputnode,  fslroi_ref,   [('in_file', 'in_file')]),
        (inputnode,  fslroi_moving,   [('in_file', 'in_file')]),
        (fslroi_moving, split_moving,   [('roi_file', 'in_file')]),
        (fslroi_ref, bet_ref, [('roi_file', 'in_file')]),
        (bet_ref, dilate, [('mask_file', 'in_file')]),

        (dilate, flirt, [('out_file', 'ref_weight'),
                        ('out_file', 'in_weight')]),
        (fslroi_ref, flirt, [('roi_file', 'reference')]),
        (split_moving, flirt, [('out_files', 'in_file')]),

        # (split_moving, list_4_flirt,[('out_files', 'in_files')]),
        # (list_4_flirt, flirt, [('out_file', 'in_file')]),

        (flirt, thres, [('out_file', 'in_file')]),
        (thres, merge, [('out_file', 'in_files')]),
        (merge, insert_ref, [('merged_file', 'in_file2')]),
        (fslroi_ref, insert_ref, [('roi_file', 'in_file1')]),
        (insert_ref, outputnode, [('out_file', 'out_file')]),
        (flirt, outputnode, [('out_matrix_file', 'out_xfms')])
    ])
`
The erros is: 

> TraitError: The 'in_file' trait of a FLIRTInputSpec instance must be an existing file name, but a value of [['/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0000.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0001.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0002.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0003.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0004.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0005.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0006.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving0/vol0007.nii.gz'], ['/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0000.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0001.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0002.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0003.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0004.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0005.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0006.nii.gz', '/Users/junhao.wen/test/DwiPreprocessingPhaseDifferenceFieldmap2/Diffusion_preprocessing_phasediff_fmb/prepare_data/b0_co_registration/split_b0_moving/mapflow/_split_b0_moving1/vol0007.nii.gz']] <type 'list'> was specified.

Actually, to parallelize the diffusion preprocessing, I tried two solutions, the first one is to really parallelize every steps like what I mentioned here(This bug), the second solution is to create datagrabber node to distribute subjects to the one-subject-one-time preprocessing pipeine(The problem for iterables tangling problems, like #2044 )

Hope to hear from you :)

Hao

**Comment by oesteban** [comment]
@anbai106 The `in_file` input of the `flirt` node is a `list`, however it is not set among in the `iterfields` of that `MapNode`.

**Comment by mgxd** [comment]
@anbai106 feel free to open if you are still running into problems
